amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+    /// }
+    ///
+    /// fn other_function() -> Result<(), Error> {
+    ///     Ok(failing_function().with_context(|_| Error::from_string("other"))?)
+    /// }
+    ///
+    /// let e = other_function().expect_err("no error");
+    ///
+    /// assert_eq!("other", e.to_string());
+    /// assert_eq!("failing", e.source().expect("no source").to_string());
+    /// assert!(e.source().expect("no source").source().is_none());
+    /// ```
+    fn with_context<C, D>(self, chain: C) -> Result<T, Error>
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>;
+}
+
+impl<T, E> ResultExt<T> for result::Result<T, E>
+where
+    E: Into<Error>,
+{
+    fn with_context<C, D>(self, chain: C) -> Result<T, Error>
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>,
+    {
+        match self {
+            Err(e) => {
+                let e = e.into();
+                Err(chain(&e).into().with_source(e))
+            }
+            Ok(value) => Ok(value),
+        }
+    }
+}
+
+/// An iterator over all the causes for this error.
+///
+/// Created using [`Error::causes`](Error::causes).
+#[derive(Debug, Clone)]
+pub struct Causes<'a> {
+    current: Option<&'a Error>,
+}
+
+impl<'a> Iterator for Causes<'a> {
+    type Item = &'a Error;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(e) = self.current {
+            self.current = e.source();
+            return Some(e);
+        }
+
+        None
+    }
+}
+
+/// Constructs an `Error` using the standard string interpolation syntax.
+///
+/// ```rust
+/// #[macro_use] extern crate amethyst_error;
+///
+/// fn main() {
+///     let err = format_err!("number: {}", 42);
+///     assert_eq!("number: 42", err.to_string());
+/// }
+/// ```
+#[macro_export]
+macro_rules! format_err {
+    ($($arg:tt)*) => { $crate::Error::from_string(format!($($arg)*)) }
+}
+
+/// Constructs an [`Error`](Error) from a string.
+pub fn err_msg<M>(message: M) -> Error
+where
+    M: 'static + Send + Sync + fmt::Debug + fmt::Display,
+{
+    Error::new(ErrMsgError { message })
+}
+
+/// Treat something that can be displayed as an error.
+struct ErrMsgError<M> {
+    message: M,
+}
+
+impl<M> error::Error for ErrMsgError<M> where M: 'static + Send + Sync + fmt::Debug + fmt::Display {}
+
+impl<M> fmt::Display for ErrMsgError<M>
+where
+    M: fmt::Display,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+impl<M> fmt::Debug for ErrMsgError<M>
+where
+    M: fmt::Debug,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+/// Test if backtracing is enabled.
+fn is_backtrace_enabled<F: Fn(&str) -> Option<ffi::OsString>>(get_var: F) -> bool {
+    match get_var(RUST_BACKTRACE) {
+        Some(ref val) if val != "0" => true,
+        _ => false,
+    }
+}
+
+static BACKTRACE_ENABLED: atomic::AtomicUsize = atomic::ATOMIC_USIZE_INIT;
```

Comment:
Could you rename this?

```suggestion
// 0: unchecked
// 1: disabled
// 2: enabled
static BACKTRACE_STATUS: atomic::AtomicUsize = atomic::ATOMIC_USIZE_INIT;
```